runtime/js/40_write_file.js

```diff
@@ -1,85 +1,53 @@
 // Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
 "use strict";
 ((window) => {
-  const { stat, statSync, chmod, chmodSync } = window.__bootstrap.fs;
-  const { open, openSync } = window.__bootstrap.files;
-  const { build } = window.__bootstrap.build;
-  const {
-    TypedArrayPrototypeSubarray,
-  } = window.__bootstrap.primordials;
+  const core = window.__bootstrap.core;
+  const { pathFromURL } = window.__bootstrap.util;
 
   function writeFileSync(
     path,
     data,
     options = {},
   ) {
     options.signal?.throwIfAborted();
-    if (options.create !== undefined) {
-      const create = !!options.create;
-      if (!create) {
-        // verify that file exists
-        statSync(path);
-      }
-    }
-
-    const openOptions = options.append
-      ? { write: true, create: true, append: true }
-      : { write: true, create: true, truncate: true };
-    const file = openSync(path, openOptions);
-
-    if (
-      options.mode !== undefined &&
-      options.mode !== null &&
-      build.os !== "windows"
-    ) {
-      chmodSync(path, options.mode);
-    }
-
-    let nwritten = 0;
-    while (nwritten < data.length) {
-      nwritten += file.writeSync(TypedArrayPrototypeSubarray(data, nwritten));
-    }
-
-    file.close();
+    core.opSync("op_write_file_sync", {
+      path: pathFromURL(path),
+      data,
+      mode: options.mode,
+      append: options.append ?? false,
+      create: options.create ?? true,
+    });
   }
 
   async function writeFile(
     path,
     data,
     options = {},
   ) {
-    if (options.create !== undefined) {
-      const create = !!options.create;
-      if (!create) {
-        // verify that file exists
-        await stat(path);
-      }
-    }
-
-    const openOptions = options.append
-      ? { write: true, create: true, append: true }
-      : { write: true, create: true, truncate: true };
-    const file = await open(path, openOptions);
-
-    if (
-      options.mode !== undefined &&
-      options.mode !== null &&
-      build.os !== "windows"
-    ) {
-      await chmod(path, options.mode);
+    let cancelRid;
+    let abortHandler;
+    if (options.signal) {
+      options.signal.throwIfAborted();
+      cancelRid = core.opSync("op_cancel_handle");
+      abortHandler = options.signal
+        .addEventListener("abort", () => core.tryClose(cancelRid));
```

Comment:
Instead of using an event listener, it should register an internal "abort step" on an `AbortSignal`. The event could be cancelled by the user, and is not synchronous.